components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java

```diff
@@ -1783,134 +1783,362 @@ private int handleStartIndexEqualsNULL(Integer startIndex) {
+                        attributeValue);
+            } else {
+                userList = getUserNames(SCIMConstants.CommonSchemaConstants.ID_URI, filterOperation, attributeValue);
+            }
+            // Get the roles of the users.
+            Set<String> fullRoleList = new HashSet<>();
+            if (userList != null) {
+                for (String userName : userList) {
+                    fullRoleList.addAll(Arrays.asList(carbonUM.getRoleListOfUser(userName)));
+                }
+            }
+            return fullRoleList.toArray(new String[0]);
+        } else if (attributeName.equals(SCIMConstants.GroupSchemaConstants.DISPLAY_NAME_URI)) {
+            return getRoleNames(filterOperation, attributeValue);
+        } else {
+            try {
+                return getGroupNamesFromDB(attributeName, filterOperation, attributeValue);
+            } catch (IdentitySCIMException e) {
+                throw new CharonException("Error in retrieving SCIM Group information from database.", e);
+            }
+        }
+    }
+
+    /**
+     * Method to append domain name in front af the attribute value to perform filter in a specific domain.
+     *
+     * @param domainName
+     * @param node
+     */
+    private String updateAttributeValue(String domainName, ExpressionNode node) throws BadRequestException {
+
+        String attributeName = node.getAttributeValue();
+        String filterOperation = node.getOperation();
+        String attributeValue = node.getValue();
+        String extractedDomain;
+
+        if (StringUtils.isNotEmpty(domainName)) {
+            if (isDomainNameEmbeddedInGroupAttributeValue(filterOperation, attributeName, attributeValue)) {
+                int indexOfDomainSeparator = attributeValue.indexOf(CarbonConstants.DOMAIN_SEPARATOR);
+                extractedDomain = attributeValue.substring(0, indexOfDomainSeparator).toUpperCase();
+
+                // Check whether the domain name is equal to the extracted domain name from attribute value.
+                if (StringUtils.isNotEmpty(extractedDomain) && !extractedDomain.equalsIgnoreCase(domainName)) {
+                    throw new BadRequestException(String.format(
+                            " Domain name: %s in the domain parameter does not match with the domain name: %s in "
+                                    + "search attribute value of %s claim.", domainName, extractedDomain,
+                            attributeName));
+                }
+
+                if (StringUtils.isEmpty(extractedDomain)) {
+                    return domainName + CarbonConstants.DOMAIN_SEPARATOR + attributeValue;
+                } else {
+                    return attributeValue;
+                }
+            } else {
+                return domainName + CarbonConstants.DOMAIN_SEPARATOR + attributeValue;
+            }
+        } else {
+            return attributeValue;
```

Comment:
the case when the domain param is not there, but attribute value contains the domain, how we proceed?

Reply:
If the attribute value contains a domain it is returned without checking. So no change to the original attribute value.